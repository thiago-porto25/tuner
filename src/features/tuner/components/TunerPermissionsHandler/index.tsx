import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { Linking, PermissionsAndroid } from 'react-native';

import RNExitApp from 'react-native-exit-app';
import { ButtonProps } from 'react-native-paper';

import testID from '@/features/tuner/constants/testIDs.constants';
import { en, pt } from '@/features/tuner/i18n/supportedLanguages';
import { TunerPermissionsHandlerProps } from '@/features/tuner/types/tunerPermissionsHandlerProps.interface';
import ActionModal from '@/shared/components/ActionModal';
import checkMultiplePermissions from '@/shared/utils/checkMultiplePermissions.util';
import getI18n from '@/shared/utils/getI18n.util';
import requestPermissions from '@/shared/utils/requestPermissions.util';

const i18n = getI18n({ en, pt });
const permissions = [PermissionsAndroid.PERMISSIONS.RECORD_AUDIO];

function TunerPermissionsHandler({
  setArePermissionsGranted,
}: TunerPermissionsHandlerProps) {
  const [isPermissionErrorModalVisible, setIsPermissionErrorModalVisible] =
    useState(false);
  const [isPermissionDeniedModalVisible, setIsPermissionDeniedModalVisible] =
    useState(false);
  const [arePermissionsNeverAskAgain, setArePermissionsNeverAskAgain] =
    useState(false);

  function handlePermissionsDenied(): void {
    setIsPermissionErrorModalVisible(false);
    setIsPermissionDeniedModalVisible(true);
  }

  function handlePermissionsError(): void {
    setIsPermissionDeniedModalVisible(false);
    setIsPermissionErrorModalVisible(true);
  }

  function handlePermissionsNeverAskAgain(): void {
    setArePermissionsNeverAskAgain(true);
    setIsPermissionDeniedModalVisible(true);
    setIsPermissionErrorModalVisible(false);
  }

  function leaveApp(): void {
    RNExitApp.exitApp();
  }

  const handlePermissionsSuccess = useCallback(() => {
    setIsPermissionDeniedModalVisible(false);
    setIsPermissionErrorModalVisible(false);
    setArePermissionsNeverAskAgain(false);
    setArePermissionsGranted(true);
  }, [setArePermissionsGranted]);

  const requestMandatoryPermissions = useCallback(async () => {
    await requestPermissions(
      permissions,
      handlePermissionsSuccess,
      handlePermissionsDenied,
      handlePermissionsError,
      handlePermissionsNeverAskAgain,
    );
  }, [handlePermissionsSuccess]);

  const retryDeniedPermissions = useCallback(async () => {
    setIsPermissionDeniedModalVisible(false);
    setIsPermissionErrorModalVisible(false);

    if (!arePermissionsNeverAskAgain) {
      await requestMandatoryPermissions();
      return;
    }

    Linking.openSettings();
  }, [arePermissionsNeverAskAgain, requestMandatoryPermissions]);

  const retryErroredPermissions = useCallback(async () => {
    setIsPermissionErrorModalVisible(false);
    await requestMandatoryPermissions();
  }, [requestMandatoryPermissions]);

  useEffect(() => {
    async function handlePermissions(): Promise<void> {
      if (await checkMultiplePermissions(permissions)) {
        handlePermissionsSuccess();
        return;
      }

      await requestMandatoryPermissions();
    }

    handlePermissions();
  }, [handlePermissionsSuccess, requestMandatoryPermissions]);

  const deniedModalButtonsContent = useMemo(
    (): ButtonProps[] => [
      {
        mode: 'outlined',
        onPress: leaveApp,
        children: i18n.t('tunerScreen.permissions.denied.leaveButtonLabel'),
        testID: arePermissionsNeverAskAgain
          ? testID.TUNER_NEVER_ASK_PERMISSION_MODAL_LEAVE_BUTTON
          : testID.TUNER_DENIED_PERMISSION_MODAL_LEAVE_BUTTON,
        key: 1,
      },
      {
        mode: 'contained',
        onPress: retryDeniedPermissions,
        children: i18n.t(
          arePermissionsNeverAskAgain
            ? 'tunerScreen.permissions.neverAsk.retryButtonLabel'
            : 'tunerScreen.permissions.denied.retryButtonLabel',
        ),
        testID: arePermissionsNeverAskAgain
          ? testID.TUNER_NEVER_ASK_PERMISSION_MODAL_RETRY_BUTTON
          : testID.TUNER_DENIED_PERMISSION_MODAL_RETRY_BUTTON,
        key: 2,
      },
    ],
    [arePermissionsNeverAskAgain, retryDeniedPermissions],
  );

  const errorModalButtonsContent = useMemo(
    (): ButtonProps[] => [
      {
        mode: 'contained',
        onPress: retryErroredPermissions,
        children: i18n.t('tunerScreen.permissions.error.retryButtonLabel'),
        testID: testID.TUNER_ERROR_PERMISSION_MODAL_RETRY_BUTTON,
        key: 1,
      },
    ],
    [retryErroredPermissions],
  );

  return (
    <>
      <ActionModal
        buttons={deniedModalButtonsContent}
        description={i18n.t(
          arePermissionsNeverAskAgain
            ? 'tunerScreen.permissions.neverAsk.description'
            : 'tunerScreen.permissions.denied.description',
        )}
        dismissableBackButton={false}
        overlayAccessibilityLabel={i18n.t(
          'tunerScreen.permissions.denied.overlayLabel',
        )}
        testID={testID.TUNER_DENIED_PERMISSION_MODAL}
        title={i18n.t('tunerScreen.permissions.denied.title')}
        visible={isPermissionDeniedModalVisible}
      />

      <ActionModal
        buttons={errorModalButtonsContent}
        description={i18n.t('tunerScreen.permissions.error.description')}
        dismissable={false}
        dismissableBackButton={false}
        overlayAccessibilityLabel={i18n.t(
          'tunerScreen.permissions.error.overlayLabel',
        )}
        testID={testID.TUNER_ERROR_PERMISSION_MODAL}
        title={i18n.t('tunerScreen.permissions.error.title')}
        visible={isPermissionErrorModalVisible}
      />
    </>
  );
}

export default TunerPermissionsHandler;
