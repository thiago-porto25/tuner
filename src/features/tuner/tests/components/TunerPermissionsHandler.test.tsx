import React from 'react';
import {
  Linking,
  Permission,
  PermissionsAndroid,
  PermissionStatus,
} from 'react-native';

import MockDate from 'mockdate';
import RNExitApp from 'react-native-exit-app';

import { act, fireEvent, render, screen } from '@/__tests__/test-utils';
import TunerPermissionsHandler from '@/features/tuner/components/TunerPermissionsHandler';
import testID from '@/features/tuner/constants/testIDs.constants';

const audioPermission = PermissionsAndroid.PERMISSIONS.RECORD_AUDIO;

jest.mock(
  'react-native//Libraries/PermissionsAndroid/PermissionsAndroid',
  () => ({
    check: jest.fn().mockResolvedValue(false),
    requestMultiple: jest.fn().mockResolvedValue({
      'android.permission.RECORD_AUDIO': 'granted',
    }),
    RESULTS: {
      GRANTED: 'granted',
      DENIED: 'denied',
      NEVER_ASK_AGAIN: 'never_ask_again',
    },
    PERMISSIONS: {
      RECORD_AUDIO: 'android.permission.RECORD_AUDIO',
    },
  }),
);

jest.mock('react-native//Libraries/Linking/Linking');

jest.mock('react-native-exit-app', () => ({
  exitApp: jest.fn(),
}));

describe('TunerPermissionsHandler', () => {
  const mockedPermissionsAndroid = PermissionsAndroid as jest.Mocked<
    typeof PermissionsAndroid
  >;
  const mockedLinking = Linking as jest.Mocked<typeof Linking>;

  beforeAll(() => {
    jest.useFakeTimers({ legacyFakeTimers: true });
    MockDate.set(0);
  });

  afterEach(() => {
    mockedPermissionsAndroid.requestMultiple.mockReset();
    mockedPermissionsAndroid.check.mockReset();
    mockedLinking.openSettings.mockReset();
  });

  it('shows no modals if permissions are already accepted', async () => {
    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.DENIED,
    } as Required<{ [key in Permission]: PermissionStatus }>);
    mockedPermissionsAndroid.check.mockResolvedValueOnce(true);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    await act(async () => {
      global.timeTravel(500);
    });

    expect(
      screen.queryByTestId(testID.TUNER_DENIED_PERMISSION_MODAL),
    ).toBeNull();
    expect(
      screen.queryByTestId(
        testID.TUNER_NEVER_ASK_PERMISSION_MODAL_RETRY_BUTTON,
      ),
    ).toBeNull();
    expect(
      screen.queryByTestId(testID.TUNER_ERROR_PERMISSION_MODAL),
    ).toBeNull();
  });

  it('shows denied permission modal when the user denies permissions', async () => {
    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.DENIED,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    expect(
      await screen.findByTestId(testID.TUNER_DENIED_PERMISSION_MODAL),
    ).toBeTruthy();
  });

  it('shows permission error modal when an error occurs while requesting permissions', async () => {
    mockedPermissionsAndroid.requestMultiple.mockRejectedValueOnce(null);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    expect(
      await screen.findByTestId(testID.TUNER_ERROR_PERMISSION_MODAL),
    ).toBeTruthy();
  });

  it('shows never ask again permission modal when the user denies permissions more than once', async () => {
    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.DENIED,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    expect(
      await screen.findByTestId(testID.TUNER_DENIED_PERMISSION_MODAL),
    ).toBeTruthy();

    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.NEVER_ASK_AGAIN,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    fireEvent.press(
      screen.getByTestId(testID.TUNER_DENIED_PERMISSION_MODAL_RETRY_BUTTON),
    );

    expect(
      await screen.findByTestId(
        testID.TUNER_NEVER_ASK_PERMISSION_MODAL_LEAVE_BUTTON,
      ),
    ).toBeTruthy();
  });

  it('handles retry of permissions correctly when the user denies permissions once', async () => {
    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.DENIED,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    expect(
      await screen.findByTestId(testID.TUNER_DENIED_PERMISSION_MODAL),
    ).toBeTruthy();

    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.GRANTED,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    fireEvent.press(
      screen.getByTestId(testID.TUNER_DENIED_PERMISSION_MODAL_RETRY_BUTTON),
    );

    await act(async () => {
      global.timeTravel(500);
    });

    expect(
      screen.queryByTestId(testID.TUNER_DENIED_PERMISSION_MODAL),
    ).toBeNull();
  });

  it('handles retry of permissions correctly when the user denies permissions more than once', async () => {
    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.NEVER_ASK_AGAIN,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    expect(
      await screen.findByTestId(
        testID.TUNER_NEVER_ASK_PERMISSION_MODAL_LEAVE_BUTTON,
      ),
    ).toBeTruthy();

    fireEvent.press(
      screen.getByTestId(testID.TUNER_NEVER_ASK_PERMISSION_MODAL_RETRY_BUTTON),
    );

    expect(mockedLinking.openSettings).toHaveBeenCalledTimes(1);
  });

  it('handles retry of permissions correctly when an error occurs requesting permissions', async () => {
    mockedPermissionsAndroid.requestMultiple.mockRejectedValueOnce(null);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    expect(
      await screen.findByTestId(testID.TUNER_ERROR_PERMISSION_MODAL),
    ).toBeTruthy();

    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.GRANTED,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    fireEvent.press(
      screen.getByTestId(testID.TUNER_ERROR_PERMISSION_MODAL_RETRY_BUTTON),
    );

    await act(async () => {
      global.timeTravel(500);
    });

    expect(
      screen.queryByTestId(testID.TUNER_ERROR_PERMISSION_MODAL),
    ).toBeNull();
  });

  it('handles correctly press on the leave app button in denied and never ask again modals', async () => {
    mockedPermissionsAndroid.requestMultiple.mockResolvedValueOnce({
      [audioPermission]: PermissionsAndroid.RESULTS.NEVER_ASK_AGAIN,
    } as Required<{ [key in Permission]: PermissionStatus }>);

    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    expect(
      await screen.findByTestId(
        testID.TUNER_NEVER_ASK_PERMISSION_MODAL_RETRY_BUTTON,
      ),
    ).toBeTruthy();

    fireEvent.press(
      screen.getByTestId(testID.TUNER_NEVER_ASK_PERMISSION_MODAL_LEAVE_BUTTON),
    );
    expect(RNExitApp.exitApp).toHaveBeenCalledTimes(1);
  });

  it('matches the snapshot', async () => {
    render(<TunerPermissionsHandler setArePermissionsGranted={jest.fn} />);

    await act(async () => {
      global.timeTravel(500);
    });

    expect(screen.toJSON()).toMatchSnapshot();
  });
});
