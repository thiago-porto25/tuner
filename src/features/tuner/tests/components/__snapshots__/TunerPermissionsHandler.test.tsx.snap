// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TunerPermissionsHandler matches the snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  />
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[Function]}
      pointerEvents="auto"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="Tuner-ErrorPermissionModal"
    >
      <View
        accessibilityLabel="tunerScreen.permissions.error.overlayLabel"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        importantForAccessibility="no"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "rgba(50, 47, 55, 0.4)",
            "flex": 1,
            "opacity": 0,
          }
        }
        testID="Tuner-ErrorPermissionModal-backdrop"
      />
      <View
        pointerEvents="box-none"
        style={
          [
            {
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "marginBottom": 0,
              "marginTop": 0,
            },
            undefined,
          ]
        }
        testID="Tuner-ErrorPermissionModal-wrapper"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.15,
              "shadowRadius": 3,
            }
          }
          testID="Tuner-ErrorPermissionModal-surface-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "transparent",
                "flex": undefined,
                "justifyContent": "center",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 1,
              }
            }
            testID="Tuner-ErrorPermissionModal-surface"
          >
            <View
              style={
                {
                  "backgroundColor": "rgba(255, 251, 254, 1)",
                  "borderTopLeftRadius": 12,
                  "borderTopRightRadius": 12,
                  "justifyContent": "space-between",
                  "minHeight": 256,
                  "paddingBottom": 24,
                  "paddingLeft": 24,
                  "paddingRight": 24,
                  "paddingTop": 24,
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  {
                    "marginBottom": 24,
                    "marginTop": 16,
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "Inter-Regular",
                          "fontSize": 22,
                          "fontWeight": "400",
                          "letterSpacing": 0,
                          "lineHeight": 28,
                        },
                        {
                          "fontFamily": "Inter-Bold",
                          "marginBottom": 8,
                        },
                      ],
                    ]
                  }
                >
                  tunerScreen.permissions.error.title
                </Text>
                <View
                  style={{}}
                >
                  <Text
                    style={
                      [
                        {
                          "textAlign": "left",
                        },
                        {
                          "color": "rgba(28, 27, 31, 1)",
                          "fontFamily": "Inter-Regular",
                          "fontWeight": "400",
                          "letterSpacing": 0,
                        },
                        {
                          "writingDirection": "ltr",
                        },
                        {},
                      ]
                    }
                  >
                    tunerScreen.permissions.error.description
                  </Text>
                </View>
              </View>
              <View
                style={
                  {
                    "gap": 8,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "#f48225",
                      "borderRadius": 20,
                      "shadowColor": "#000",
                      "shadowOffset": {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                  testID="Tuner-ErrorPermissionModal-RetryButton-container-outer-layer"
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "backgroundColor": "#f48225",
                        "borderColor": "transparent",
                        "borderRadius": 20,
                        "borderStyle": "solid",
                        "borderWidth": 0,
                        "flex": undefined,
                        "minWidth": 64,
                        "shadowColor": "#000",
                        "shadowOffset": {
                          "height": 0,
                          "width": 0,
                        },
                        "shadowOpacity": 0,
                        "shadowRadius": 0,
                      }
                    }
                    testID="Tuner-ErrorPermissionModal-RetryButton-container"
                  >
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        {
                          "busy": undefined,
                          "checked": undefined,
                          "disabled": false,
                          "expanded": undefined,
                          "selected": undefined,
                        }
                      }
                      accessibilityValue={
                        {
                          "max": undefined,
                          "min": undefined,
                          "now": undefined,
                          "text": undefined,
                        }
                      }
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        [
                          {
                            "overflow": "hidden",
                          },
                          {
                            "borderRadius": 20,
                          },
                        ]
                      }
                      testID="Tuner-ErrorPermissionModal-RetryButton"
                    >
                      <View
                        style={
                          [
                            {
                              "alignItems": "center",
                              "flexDirection": "row",
                              "justifyContent": "center",
                            },
                            undefined,
                          ]
                        }
                      >
                        <Text
                          numberOfLines={1}
                          selectable={false}
                          style={
                            [
                              {
                                "textAlign": "left",
                              },
                              {
                                "color": "rgba(28, 27, 31, 1)",
                                "writingDirection": "ltr",
                              },
                              [
                                {
                                  "fontFamily": "Inter-Medium",
                                  "fontSize": 14,
                                  "fontWeight": "400",
                                  "letterSpacing": 0.1,
                                  "lineHeight": 20,
                                },
                                [
                                  {
                                    "marginHorizontal": 16,
                                    "marginVertical": 9,
                                    "textAlign": "center",
                                  },
                                  false,
                                  {
                                    "marginHorizontal": 24,
                                    "marginVertical": 10,
                                  },
                                  undefined,
                                  false,
                                  {
                                    "color": "rgba(255, 255, 255, 1)",
                                    "fontFamily": "Inter-Medium",
                                    "fontSize": 14,
                                    "fontWeight": "400",
                                    "letterSpacing": 0.1,
                                    "lineHeight": 20,
                                  },
                                  undefined,
                                ],
                              ],
                            ]
                          }
                          testID="Tuner-ErrorPermissionModal-RetryButton-text"
                        >
                          tunerScreen.permissions.error.retryButtonLabel
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</RNCSafeAreaProvider>
`;
